firestore/src/common/local_cache_settings.cc

```diff
@@ -0,0 +1,207 @@
+
+MemoryLruGCSettings MemoryLruGCSettings::WithSizeBytes(int64_t size) {
+  return {MemoryLruGCSettingsInternal{
+      settings_internal_->core_settings().WithSizeBytes(size)}};
+}
+
+MemoryCacheSettings MemoryCacheSettings::Create() { return {}; }
+
+MemoryCacheSettings MemoryCacheSettings::CreateFromCoreSettings(
+    const CoreMemorySettings& core_settings) {
+  auto result = MemoryCacheSettings{};
+  result.settings_internal_ =
+      std::make_unique<MemoryCacheSettingsInternal>(core_settings);
+  return result;
+}
+
+MemoryCacheSettings::MemoryCacheSettings() {
+  settings_internal_ =
+      std::make_unique<MemoryCacheSettingsInternal>(CoreMemorySettings{});
+}
+
+MemoryCacheSettings::MemoryCacheSettings(const MemoryCacheSettings& other) {
+  settings_internal_ =
+      std::make_unique<MemoryCacheSettingsInternal>(*other.settings_internal_);
+}
+
+MemoryCacheSettings& MemoryCacheSettings::operator=(
+    const MemoryCacheSettings& other) {
+  if (this == &other) {
+    return *this;
+  }
+
+  settings_internal_ =
+      std::make_unique<MemoryCacheSettingsInternal>(*other.settings_internal_);
+  return *this;
+}
+
+MemoryCacheSettings::~MemoryCacheSettings() { settings_internal_.reset(); }
+
+MemoryCacheSettings MemoryCacheSettings::WithGarbageCollectorSettings(
+    const MemoryGarbageCollectorSettings& settings) const {
+  MemoryCacheSettings result{*this};
+  CoreMemorySettings core_settings = result.settings_internal_->core_settings();
+  result.settings_internal_->set_core_settings(
+      core_settings.WithMemoryGarbageCollectorSettings(
+          settings.core_gc_settings()));
+  return result;
+}
+
+const CoreCacheSettings& MemoryCacheSettings::core_cache_settings() const {
+  return settings_internal_->core_settings();
+}
+
+bool operator==(const MemoryCacheSettings& lhs,
+                const MemoryCacheSettings& rhs) {
+  return &lhs == &rhs || (*lhs.settings_internal_ == *rhs.settings_internal_);
+}
+
+bool operator!=(const MemoryCacheSettings& lhs,
+                const MemoryCacheSettings& rhs) {
+  return !(lhs == rhs);
+}
+
+bool operator==(const PersistentCacheSettings& lhs,
+                const PersistentCacheSettings& rhs) {
+  return &lhs == &rhs || (*lhs.settings_internal_ == *rhs.settings_internal_);
+}
+
+bool operator!=(const PersistentCacheSettings& lhs,
+                const PersistentCacheSettings& rhs) {
+  return !(lhs == rhs);
+}
+
+bool operator==(const MemoryEagerGCSettings& lhs,
+                const MemoryEagerGCSettings& rhs) {
+  return &lhs == &rhs || (*lhs.settings_internal_ == *rhs.settings_internal_);
+}
+
+bool operator!=(const MemoryEagerGCSettings& lhs,
+                const MemoryEagerGCSettings& rhs) {
+  return !(lhs == rhs);
+}
+
+bool operator==(const MemoryLruGCSettings& lhs,
+                const MemoryLruGCSettings& rhs) {
+  return &lhs == &rhs || (*lhs.settings_internal_ == *rhs.settings_internal_);
+}
+
+bool operator!=(const MemoryLruGCSettings& lhs,
```

Comment:
Conventionally, the `operator!=` is defined in the .h file as "inline". Consider doing that here too.